packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,418 @@
+    }
+
+    // Natively, the attributes observed by the registered definition are going to be taken
+    // care of by the browser, only the difference between the two sets has to be taken
+    // care by the patched version.
+    return new Set(
+        [...instanceDefinition.observedAttributes].filter(
+            (x) => !registeredDefinition.observedAttributes.has(x)
+        )
+    );
+}
+
+// Helper to patch `setAttribute`/`getAttribute` to implement `attributeChangedCallback`.
+// Why is this necessary? Well basically, you can't change the `observedAttributes` after
+// a custom element is defined. So with pivots, if two classes share the same tag name,
+// and the second class observes attributes that aren't observed by the first one,
+// then those attributes can never be observed by the native `observedAttributes` system.
+// So we have to simulate it by patching `getAttribute`/`removeAttribute`. Note that
+// we only do this when absolutely necessary, though; i.e. because we've determined
+// that we aren't observing the attributes we need to.
+function patchAttributes(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instanceDefinition: Definition
+) {
+    const newObservedAttributes = getNewObservedAttributes(
+        registeredDefinition,
+        instanceDefinition
+    );
+    if (getNewObservedAttributes(registeredDefinition, instanceDefinition).size === 0) {
+        return;
+    }
+    const { attributeChangedCallback } = instanceDefinition;
+
+    // Patch the instance.
+    // Note we use the native `getAttribute` rather than the super's `getAttribute` because
+    // we don't actually want it to be observable that we're calling `getAttribute` from
+    // `setAttribute` and `removeAttribute`.
+    defineProperties(instance, {
+        setAttribute: {
+            value: function setAttribute(name: string, value: any) {
+                if (newObservedAttributes.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    nativeSetAttribute.call(this, name, value);
+                    attributeChangedCallback!.call(this, name, old, value + '');
+                } else {
+                    nativeSetAttribute.call(this, name, value);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+        removeAttribute: {
+            value: function removeAttribute(name: string) {
+                if (newObservedAttributes.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    nativeRemoveAttribute.call(this, name);
+                    attributeChangedCallback!.call(this, name, old, null);
+                } else {
+                    nativeRemoveAttribute.call(this, name);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+    });
```

Comment:
FWIW @pmdartus since we were talking about this earlier: I checked using vanilla WCs, and it seems that the `attributeChangedCallback` is actually invoked synchronously:

![Screen Shot 2022-06-15 at 3 13 13 PM](https://user-images.githubusercontent.com/283842/173941129-a57e9ac3-a121-48f2-a925-a7d0ef0c1a83.png)

Reply:
You are correct, custom element lifecycle callbacks are invoked synchronously from the user script standpoint. 

The HTML specification defines [custom element reactions](https://html.spec.whatwg.org/#custom-element-reactions). Custom element reactions are an umbrella term for all the CE lifecycle hooks. Custom elements reactions are quite complex, here is a relevant paragraph in this context.

> The way in which [custom element reactions](https://html.spec.whatwg.org/#concept-custom-element-reaction) are invoked is done with special care, to avoid running author code during the middle of delicate operations. Effectively, they are delayed until "just before returning to user script". This means that for most purposes they appear to execute synchronously, but in the case of complicated composite operations (like [cloning](https://dom.spec.whatwg.org/#concept-node-clone), or [range](https://dom.spec.whatwg.org/#concept-range) manipulation), they will instead be delayed until after all the relevant user agent processing steps have completed, and then run together as a batch.

So even if from the user script perspective those operations happen synchronously, it's not the case internally. That's what I tried to poorly explain yesterday.

One of the side effects of this approach is that it's not possible to catch errors that are thrown within custom element reactions. In the following example, the `console.error` will never evaluate:

```js
customElements.define('x-foo', class extends HTMLElement {
  static observedAttributes = ['foo'];
  attributeChangedCallback(name) {
    throw new Error('Throwing in attributeChangedCallback ' + name)
  }
})

function setAttribute() {
	const elm = document.createElement('x-foo');
  try {
    elm.setAttribute('foo', 'test');
  } catch (e) {
    console.error('!!!', e)
  }
}


async function setAttributeAsync() {
	const elm = document.createElement('x-foo');
  elm.setAttribute('foo', 'test');
}


setAttribute();

setAttributeAsync().catch(err => {
	console.error('!!!!', err);
})
```

It will be an observable change with this patch since I don't it is possible to emulate this in userland. But I guess it's fine since it's a pretty edge case.

Reply:
I suppose we could emulate this by doing a try/catch when we call `attributeChangedCallback`? Something like:

```js
try {
  attributeChangedCallback(name, oldVal, newVal)
} catch (err) {
  // Presumably logs in the DevTools similar to an uncaught error in attributeChangedCallback?
  Promise.resolve().then(() => { 
    throw err
  })
}
```

Reply:
This is done now and I added tests for it.